lib/backend/k8s/resources/workloadendpoint.go

```diff
@@ -0,0 +1,192 @@
+	log.Warn("Operation Watch is not supported on WorkloadEndpoint type")
+	return nil, cerrors.ErrorOperationNotSupported{
+		Identifier: list,
+		Operation:  "Watch",
+	}
+}
+
+/*
+
+// applyWorkloadEndpoint patches the existing Pod to include an IP address, if
+// one has been set on the workload endpoint.
+// TODO: This is only required as a workaround for an upstream k8s issue.  Once fixed,
+// this should be a no-op. See https://github.com/kubernetes/kubernetes/issues/39113
+func (c *KubeClient) applyWorkloadEndpoint(k *model.KVPair) (*model.KVPair, error) {
+	ips := k.Value.(*model.WorkloadEndpoint).IPv4Nets
+	if len(ips) > 0 {
+		log.Debugf("Applying workload with IPs: %+v", ips)
+		ns, name := c.converter.ParseWorkloadID(k.Key.(model.WorkloadEndpointKey).WorkloadID)
+		pod, err := c.clientSet.Pods(ns).Get(name, metav1.GetOptions{})
+		if err != nil {
+			return nil, resources.K8sErrorToCalico(err, k.Key)
+		}
+		pod.Status.PodIP = ips[0].IP.String()
+		pod, err = c.clientSet.Pods(ns).UpdateStatus(pod)
+		if err != nil {
+			return nil, resources.K8sErrorToCalico(err, k.Key)
+		}
+		log.Debugf("Successfully applied pod: %+v", pod)
+		return c.converter.PodToWorkloadEndpoint(pod)
+	}
+	return k, nil
+}
+
+// listWorkloadEndpoints lists WorkloadEndpoints from the k8s API based on existing Pods.
+func (c *KubeClient) listWorkloadEndpoints(ctx context.Context, l model.ResourceListOptions, revision string) (*model.KVPairList, error) {
+	// If a workload is provided, we can do an exact lookup of this
+	// workload endpoint.
+	if l.Name != "" {
+		kvp, err := c.getWorkloadEndpoint(ctx, model.ResourceKey{
+			Name: l.Name,
+			Kind: l.Kind,
+		}, revision)
+		if err != nil {
+			switch err.(type) {
+			// Return empty slice of KVPair if the object doesn't exist, return the error otherwise.
+			case cerrors.ErrorResourceDoesNotExist:
+				return &model.KVPairList{
+					KVPairs:  []*model.KVPair{},
+					Revision: revision,
+				}, nil
+			default:
+				return nil, err
+			}
+		}
+
+		return &model.KVPairList{
+			KVPairs:  []*model.KVPair{kvp},
+			Revision: revision,
+		}, nil
+	}
+
+	// Otherwise, enumerate all pods in all namespaces.
+	// We don't yet support hostname, orchestratorID, for the k8s backend.
+	pods, err := c.clientSet.Pods("").List(metav1.ListOptions{})
+	if err != nil {
+		return nil, resources.K8sErrorToCalico(err, l)
+	}
+
+	// For each Pod, return a workload endpoint.
+	ret := []*model.KVPair{}
+	for _, pod := range pods.Items {
+		// Decide if this pod should be displayed.
+		if !c.converter.IsReadyCalicoPod(&pod) {
+			continue
+		}
+
+		kvp, err := c.converter.PodToWorkloadEndpoint(&pod)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, kvp)
+	}
+	return &model.KVPairList{
+		KVPairs:  ret,
+		Revision: revision,
+	}, nil
+}
+
+// getWorkloadEndpoint gets the WorkloadEndpoint from the k8s API based on existing Pods.
```

Comment:
@doublek you're doing all of this in your PR?

Reply:
Yes.